compiler/src/dotty/tools/dotc/typer/Implicits.scala

```diff
@@ -1122,17 +1122,29 @@ trait Implicits:
             adapt(generated, pt.widenExpr, locked)
         else {
           def untpdGenerated = untpd.TypedSplice(generated)
-          def producesConversion(info: Type): Boolean = info match
-            case info: PolyType => producesConversion(info.resType)
-            case info: MethodType if info.isImplicitMethod => producesConversion(info.resType)
-            case _ => info.derivesFrom(defn.ConversionClass)
+          def conversionResultType(info: Type): Type = info match
+            case info: PolyType => conversionResultType(info.resType)
+            case info: MethodType if info.isImplicitMethod => conversionResultType(info.resType)
+            case _ if info.derivesFrom(defn.ConversionClass) => pt match
+              case selProto: SelectionProto =>
+                info.baseType(defn.ConversionClass) match
+                  case AppliedType(_, List(_, restpe)) if selProto.isMatchedBy(restpe) =>
+                    // if we embed the SelectionProto as the Conversion result type
+                    // it might end up within a GADT cast type
+                    // so instead replace it with the targeted conversion type, if it matches
+                    // see tests/pos/i15867.scala.
+                    restpe
+                  case _ => pt
```

Comment:
But in this case and also on the next line we still get a SelectionProto in the result type of the conversion, which looks fishy. Can we replace with NoType instead?